File: 2019/python/day04.py
input_start = 234208
input_end = 765869

def meets_rules_one(number):
    digits = [int(n) for n in list(str(number))]

    for idx in range(len(digits) - 1):
        if digits[idx + 1] < digits[idx]: return False

    for idx in range(len(digits) - 1):
        if digits[idx] == digits[idx + 1]: return True

    return False

def meets_rules_two(number):
    digits = [int(n) for n in list(str(number))]

    for idx in range(len(digits) - 1):
       if digits[idx + 1] < digits[idx]: return False

    for n in range(0, 10):
        if digits.count(n) == 2: return True

    return False


part_one = len([n for n in range(input_start, input_end + 1) if meets_rules_one(n)])

part_two = len([n for n in range(input_start, input_end + 1) if meets_rules_two(n)])

print(part_one)  # 1246

print(part_two)  # 814

File: 2019/javascript/day04.js
const input_start = 234208
const input_end = 765869

const meetsRules = (number, groupMatchFunc) => {
  const digits = String(number).split('')

  if (Number(digits.sort().join('')) !== number) return false

  for (const d of new Set(digits)) {
    const count = digits.reduce((t, n) => (n === d ? t + 1 : t), 0)
    if (groupMatchFunc(count)) return true
  }

  return false
}

const numberOfMatches = (matcher) => {
  let matches = 0

  for (n = input_start; n <= input_end; n++) {
    if (meetsRules(n, matcher)) matches++
  }

  return matches
}

const part_one = numberOfMatches((n) => n >= 2)
const part_two = numberOfMatches((n) => n == 2)

console.log(part_one) // 1246
console.log(part_two) // 814
